src/Bart/Jenkins/Job.php

```diff
@@ -9,52 +9,122 @@
-	 */
-	public function __construct($domain, $job_name)
+	 * Job constructor. Loads metadata about a project.
+	 * @param string $domain The
+	 * @param string $baseProject The top-level name of your Project.
+	 * @param array $subProjects If your Jenkins Job is not defined in the top-level project,
+	 * this parameter must be passed in. For example, if your Job is defined in the following
+	 * third level project: Base->Build->Example, you must pass in 'Base' as the $baseProject
+	 * parameter, and the array ['Build', 'Example'] as the $subProjects parameter. The order
+	 * of the array is maintained when loading the metadata, and hence it's critical.
+	 * @param string $protocol 'http' or 'https'.
+	 * @param string $user User to authenticate against.
+	 * @param string $token API token corresponding to user.
+	 * NOTE: The $user and $token must be passed in if the Jenkins instance doesn't support
+	 * anonymous connections.
+	 * @param int $port The port can be generally be determined by the $protocol. 'http' corresponds
+	 * to port 8080, while 'https' corresponds to 443. If the $port is passed in, it will override
+	 * that determination.
+     */
+	public function __construct(
+		$domain,
+		$baseProject,
+		array $subProjects = [],
+		$protocol = 'http',
+		$user = null,
+		$token = null,
+		$port = 0
+	)
 	{
-		if (!$domain)
-		{
-			throw new \Exception('Must provide a valid domain');
+		if (!$domain) {
+			throw new \InvalidArgumentException('Must provide a valid domain');
 		}
 
-		if (!$job_name)
-		{
-			throw new \Exception('Must provide a job name');
+		if (!$baseProject) {
+			throw new \InvalidArgumentException('Must provide a base job name');
+		}
+
+		$this->curlOptions = [];
+		$baseProjectEncoded = rawurlencode($baseProject);
+		$projectPath = "job/$baseProjectEncoded";
+		if ($subProjects !== []) {
+			for ($i = 0; $i < count ($subProjects); $i++) {
+				$subProjectEncoded = rawurlencode($subProjects[$i]);
+				$projectPath .= "/job/$subProjectEncoded";
+			}
+		}
+
+
+		if ($protocol !== 'http' && $protocol !== 'https') {
+			throw new \InvalidArgumentException("The protocol must only be 'http' or 'https'");
+		}
+
+		if ($user !== null || $token !== null) {
+			if ($token === null || $user === null ) {
+				throw new \InvalidArgumentException("You must specify both a user and a token");
+			}
+			$this->curlOptions[CURLOPT_USERPWD] = "{$user}:{$token}";
+		}
+
+		$this->port = $port;
+		if ($this->port === 0) {
+			if ($protocol === 'http') {
+				$this->port = 8080;
+			} else {
+				$this->port = 443;
+			}
 		}
 
 		$this->logger = Log4PHP::getLogger(__CLASS__);
 
-		$this->base_job_url = "http://$domain:8080/job/" . rawurlencode($job_name) . '/';
-		$this->logger->debug('Base uri: ' . $this->base_job_url);
+		$this->baseJobUrl = "{$protocol}://{$domain}:{$this->port}/{$projectPath}/";
+		$this->logger->debug('Base uri: ' . $this->baseJobUrl);
 
 		$this->metadata = $this->get_json(array());
```

Comment:
In retrospect, I always disliked the fact that this class made an external API call in the constructor. What do you think about moving this into a `connect()` method or something?